presto-hudi/src/main/java/com/facebook/presto/hudi/ParquetPageSourceUtil.java

```diff
@@ -0,0 +1,210 @@
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.units.DataSize;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.FileMetaData;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.io.MessageColumnIO;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.Type;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.facebook.presto.hive.CacheQuota.NO_CACHE_CONSTRAINTS;
+import static com.facebook.presto.hive.parquet.HdfsParquetDataSource.buildHdfsParquetDataSource;
+import static com.facebook.presto.hudi.HudiErrorCode.HUDI_CANNOT_OPEN_SPLIT;
+import static com.facebook.presto.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getColumnIO;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getDescriptors;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getParquetTypeByName;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.buildPredicate;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.predicateMatches;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.apache.parquet.io.ColumnIOConverter.constructField;
+
+final class ParquetPageSourceUtil
+{
+    private ParquetPageSourceUtil() {}
+
+    public static ConnectorPageSource createParquetPageSource(
+            TypeManager typeManager,
+            HdfsEnvironment hdfsEnvironment,
+            String user,
+            Configuration configuration,
+            Path path,
+            long start,
+            long length,
+            SchemaTableName tableName,
+            List<HudiColumnHandle> regularColumns,
+            boolean useParquetColumnNames,
+            boolean failOnCorruptedParquetStatistics,
+            DataSize maxReadBlockSize,
+            boolean batchReaderEnabled,
+            boolean verificationEnabled,
+            TupleDomain<HudiColumnHandle> effectivePredicate,
+            FileFormatDataSourceStats fileFormatDataSourceStats)
+    {
+        AggregatedMemoryContext systemMemoryContext = newSimpleAggregatedMemoryContext();
```

Comment:
Since a lot of the code is similar to ParquetPageSource in hive connector, we should constantly keep it in sync with future updates unless we can find a way to extract out to a common module.

Reply:
I agree. There can be a parquet page source factory class that is independent from hive connector. This might be defined in presto-parquet module. It can benefit many connectors, including hudi, iceberg and deltalake.